modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/HectorBasedMessageStoreImpl.java

```diff
@@ -470,20 +476,63 @@ public AndesMessageMetadata getMetaData(long messageId) throws AndesException {
                                                                        long firstMsgId, int count)
             throws AndesException {
 
+        Timer timer = new Timer();
+        List<AndesMessageMetadata> messagesFromQueue = new ArrayList<AndesMessageMetadata>();
+        long lastMsgId;
+
         Context context = MetricManager.timer(Level.DEBUG, MetricsConstants.GET_NEXT_MESSAGE_METADATA_FROM_QUEUE).start();
+
+        if (firstMsgId == 0) {
+            firstMsgId = ServerStartupRecoveryUtils.getStartMessageIdForWarmStartup();
+        }
+        long messageIdDifference = ServerStartupRecoveryUtils.getMessageDifferenceForWarmStartup();
+        lastMsgId = firstMsgId + messageIdDifference;
         try {
-            return HectorDataAccessHelper
+            messagesFromQueue = HectorDataAccessHelper
                     .getMessagesFromQueue(queueName,
                             HectorConstants.META_DATA_COLUMN_FAMILY,
-                            keyspace, firstMsgId, Long.MAX_VALUE,
+                            keyspace, firstMsgId, lastMsgId,
                             count, true);
+            if (messagesFromQueue.size() == 0) {
+                readingCassandraInfoLog(timer);
+                while (lastMsgId <= lastRecoveryMessageId) {
+                    long nextMsgId = lastMsgId + 1;
+                    lastMsgId = nextMsgId + messageIdDifference;
+                    messagesFromQueue = HectorDataAccessHelper
+                            .getMessagesFromQueue(queueName,
+                                    HectorConstants.META_DATA_COLUMN_FAMILY,
+                                    keyspace, nextMsgId, lastMsgId,
+                                    count, true);
+                    if (messagesFromQueue.size() > 0) {
+                        break;
```

Comment:
Even though we need to check for the scenario with messages returned less than the count in a single query.